File: app/src/main/java/com/letter/days/activity/AnniversaryActivity.java
package com.letter.days.activity;

import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.letter.days.R;
import com.letter.days.anniversary.AnniPagerTransformer;
import com.letter.days.anniversary.Anniversary;
import com.letter.days.anniversary.AnniversaryFragment;

import org.litepal.LitePal;

import java.util.ArrayList;
import java.util.List;

public class AnniversaryActivity extends AppCompatActivity {

    private ActionBar actionBar;

    private ViewPager viewPager;

    private int anniId;
    private boolean isEdited = false;

    List<Anniversary> anniversaryList;
    List<AnniversaryFragment> anniversaryFragments;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_anniversary);
        Toolbar toolbar = findViewById(R.id.Toolbar);
        setSupportActionBar(toolbar);
        this.getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);

        actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeAsUpIndicator(R.drawable.ic_back);
            actionBar.setTitle("");
        }

        anniId = getIntent().getIntExtra("anniId", 0);

        viewPager = findViewById(R.id.view_pager);

    }

    @Override
    protected void onResume() {
        super.onResume();

        anniversaryList = LitePal.findAll(Anniversary.class);
        anniversaryFragments = new ArrayList<>();
        for (Anniversary anniversary: anniversaryList) {
            anniversaryFragments.add(AnniversaryFragment.newInstance(anniversary.getId()));
        }
        AnniFragmentPagerAdapter<AnniversaryFragment> adapter =
                new AnniFragmentPagerAdapter<>(getSupportFragmentManager(), anniversaryFragments);
        viewPager.setAdapter(adapter);
        viewPager.setCurrentItem(getFragmentPositionById(anniId), true);
        viewPager.setPageTransformer(false, new AnniPagerTransformer());
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                anniId = anniversaryFragments.get(viewPager.getCurrentItem()).getAnniId();
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.anniversary_toolbar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent intent;
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isEdited) {
                    intent = new Intent();
                    setResult(RESULT_OK, intent);
                }
                finish();
                break;

            case R.id.delete:
                AlertDialog.Builder dialog = new AlertDialog.Builder(AnniversaryActivity.this);
                dialog.setMessage("确认删除这个纪念日？");
                dialog.setPositiveButton("确认", (dialogInterface, i) -> {
                        LitePal.delete(Anniversary.class, anniId);
                        Intent intent1 = new Intent();
                        setResult(RESULT_OK, intent1);
                        Toast.makeText(AnniversaryActivity.this, "已删除", Toast.LENGTH_SHORT).show();
                        finish();
                    });
                dialog.setNegativeButton("取消", (dialogInterface, i) -> {});
                dialog.show();
                break;

            case R.id.edit:
                intent = new Intent(AnniversaryActivity.this, AddItemActivity.class);
                intent.putExtra("editType", AddItemActivity.ITEM_EDIT);
                intent.putExtra("anniId", anniId);
                startActivityForResult(intent, 1);
                break;

            default:
                break;
        }
        return true;
    }

    @Override
    public void onBackPressed() {
        if (isEdited) {
            Intent intent = new Intent();
            setResult(RESULT_OK, intent);
        }
        finish();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        switch (requestCode) {
            case 1:
                if (resultCode == RESULT_OK) {
                    isEdited = true;
                }
                break;

            default:
                break;
        }
    }

    private int getFragmentPositionById(int id) {
        for (int i = 0; i < anniversaryList.size(); i++) {
            if (anniversaryList.get(i).getId() == id) {
                return i;
            }
        }
        return 0;
    }

}

File: app/src/main/java/com/letter/days/activity/AnniversaryActivity.kt
package com.letter.days.activity

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.lifecycle.*
import androidx.viewpager.widget.ViewPager
import com.afollestad.materialdialogs.MaterialDialog
import com.letter.days.R
import com.letter.days.adapter.ViewPagerAdapter
import com.letter.days.databinding.ActivityAnniversaryBinding
import com.letter.days.transformer.AnniViewPagerTransformer
import com.letter.days.viewmodel.AnniversaryViewModel
import android.content.startActivity
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.LayerDrawable
import androidx.preference.PreferenceManager
import com.blankj.utilcode.util.ImageUtils
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.FileInputStream

/**
 * 纪念日详情界面
 * @property binding ActivityAnniversaryBinding 视图绑定
 * @property model AnniversaryViewModel view model
 * @property onPageChangeListener <no name provided> 页面改变监听
 *
 * @author Letter(nevermindzzt@gmail.com)
 * @since 1.0.0
 */
class AnniversaryActivity : BaseActivity() {

    private lateinit var binding: ActivityAnniversaryBinding
    private val model by lazy {
        ViewModelProvider(this).get(AnniversaryViewModel::class.java)
    }

    private val onPageChangeListener = object : ViewPager.OnPageChangeListener {
        override fun onPageScrollStateChanged(state: Int) = Unit

        override fun onPageScrolled(
            position: Int,
            positionOffset: Float,
            positionOffsetPixels: Int
        ) {
            model.currentPosition.value = position
        }

        override fun onPageSelected(position: Int) = Unit
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAnniversaryBinding.inflate(layoutInflater)
        setContentView(binding.root)

        /* 设置Action Bar并使能home按钮 */
        setSupportActionBar(binding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        initBinding()
        initModel()
    }

    override fun onResume() {
        super.onResume()
        model.loadDays()
    }

    /**
     * 初始化视图绑定
     */
    private fun initBinding() {
        binding.viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            private var lastPosition = 0
            private var background : LayerDrawable? = null
            private var lock = false

            override fun onPageScrollStateChanged(state: Int) = Unit

            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {
                if (PreferenceManager.getDefaultSharedPreferences(this@AnniversaryActivity)
                        .getBoolean("simple_mode", true) || lock) {
                    return
                }
                MainScope().launch {
                    try {
                        if (background == null || lastPosition != position) {
                            lock = true
                            lastPosition = position
                            withContext(Dispatchers.IO) {
                                val topDrawable =
                                    if (!model.daysList.value?.get(position)?.image.isNullOrEmpty()) {
                                        BitmapDrawable(
                                            resources,
                                            ImageUtils.fastBlur(
                                                BitmapFactory.decodeStream(
                                                    FileInputStream(model.daysList.value?.get(position)?.image)),
                                                0.3f,
                                                25f
                                            )
                                        )
                                    } else {
                                        null
                                    }
                                val bottomDrawable =
                                    if (position + 1 < model.daysList.value?.size ?: 0
                                        && !model.daysList.value?.get(position + 1)?.image.isNullOrEmpty()) {
                                        BitmapDrawable(
                                            resources,
                                            ImageUtils.fastBlur(
                                                BitmapFactory.decodeStream(
                                                    FileInputStream(model.daysList.value?.get(position + 1)?.image)),
                                                0.3f,
                                                25f
                                            )
                                        )
                                    } else {
                                        null
                                    }
                                topDrawable?.alpha = 0
                                val layers = arrayOf(bottomDrawable, topDrawable)
                                background = LayerDrawable(layers)
                            }
                            withContext(Dispatchers.Main) {
                                binding.root.background = background
                            }
                            lock = false
                        }
                    } finally {
                    }
                    withContext(Dispatchers.Main) {
                        background?.getDrawable(0)?.alpha = ((positionOffset / 1f) * 255).toInt()
                        background?.getDrawable(1)?.alpha = (((1f - positionOffset)  / 1f) * 255).toInt()
                    }
                }
            }

            override fun onPageSelected(position: Int) {
                try {
                    model.freshProgress(position)
                } catch (e: Exception) {
                }
            }
        })
    }

    /**
     * 初始化view model
     */
    private fun initModel() {
        model.apply {
            fragmentList.observe(this@AnniversaryActivity) {
                binding.viewPager.apply {
                    adapter = ViewPagerAdapter(supportFragmentManager, it)
                    setCurrentItem(
                        model.currentPosition.value
                            ?: model.getFragmentPosition(intent.getIntExtra("anniId", -1)),
                        true)
                    setPageTransformer(false, AnniViewPagerTransformer())
                    addOnPageChangeListener(onPageChangeListener)
                }
            }
        }
    }

    /**
     * 创建菜单
     * @param menu Menu 菜单
     * @return Boolean 是否创建
     */
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_anniversary_toolbar, menu)
        return true
    }

    /**
     * 菜单选项选择处理
     * @param item 被选中的选项
     * @return Boolean 动作是否被处理
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.delete -> {
                MaterialDialog(this).show {
                    message(R.string.activity_anniversary_delete_dialog_message)
                    positiveButton(R.string.dialog_positive_button) {
                        model.deleteCurrent()
                        it.dismiss()
                        this@AnniversaryActivity.finish()
                    }
                    negativeButton(R.string.dialog_negative_button)
                }
            }
            R.id.edit -> {
                startActivity(AnniEditActivity::class.java) {
                    putExtra("anniId", model.getCurrentAnniversaryId())
                }
            }
            R.id.share -> {
                model.shareCurrentAnniversary()
            }
        }
        return true
    }
}
